### ISSUE TITLE
Explore running RiboViz on Eddie

### ISSUE THREAD

**Issue by mikej888** [riboviz/riboviz#59]
University of Edinburgh ECDF Linux Compute Cluster, [Eddie](https://www.ed.ac.uk/information-services/research-support/research-computing/ecdf/high-performance-computing)

* Eddie has bioinformatics software already available.
* Get access to Eddie.
* Check if Eddie nodes have the required software to run RiboViz.
* Find out what is required for any additional packages required to run RiboViz to be made available.
* Run RiboViz on a single Eddie node.
* Identify how RiboViz could be redesigned to exploit a multi-node environment.

Relates to #21, #48, #60.

**Comment by ewallace** [comment]
@FlicAnderson and I discussed this being first milestone in @XUEXUEXUE0 (Siyin Xue's) MSc Bioinformatics project. I will ask Mike Wallis (Eddie) and Alison Meynert (IGMM bioinformatics) about the bioinformatics pipelines that they already have on Eddie. We can then discuss at this week's developer meeting.

**Comment by ewallace** [comment]
Lots of useful information from a discussion on edinburgh bioinformatics slack:

@Alanocallaghan said: "You need access to the wild west nodes, the 1GB RAM limit on standard login nodes isn't enough to run nextflow." [Alison Meynert[(https://www.ed.ac.uk/profile/dr-alison-meynert) confirmed "you need to be on one of the WW nodes with a screen or tmux session to run Nextflow."

I am not sure which of the [node sizes on Eddie (EASE login required)](https://www.wiki.ed.ac.uk/display/ResearchServices/Memory+Specification) that refers to.

Alsison Meynert and her team at IGMM bioinformatics have some nextflow pipelines working on Eddie, [config scripts here](https://git.ecdf.ed.ac.uk/isaric-4c/wp5-rna-seq/wikis/home). Alison said "there are 5 different Nextflow pipelines linked to from here - two are nf-core and I've written eddie-specific configs, 1 is forked from another repository with similar idea, and 2 I wrote myself."

**Comment by kavousan** [comment]
@XUEXUEXUE0 and @ewallace to have conversation with Eddie to see if they help us put it on Eddie or if they put it on Eddie. 

The branch is nextflow-167.

**Comment by ewallace** [comment]
We talked with Mike Wallis and John Ireland today. First goal is to try running the vignette on Eddie, using an interactive node with qlogin. This will have to wait until Eddie is back up (hopefully next week)

Use IGMM nextflow instance that is configured to use gridengine. 
*Correction:* module is anaconda, then nextflow is a conda environment, `conda install nextflow`.

We can contact John Ireland at IGMM for help, John has 20% of his time dedicated to central Research Services.

Next step, @ewallace to work on setting up access and priority compute:
* contact MSc bioinformatics about setting up costings.
* then go through storage manager for my priority compute projects button.

**Comment by mikej888** [comment]
Nextflow executors, https://www.nextflow.io/docs/latest/executor.html, describe how to run Nextflow under various cluster, HPC and cloud infrastructures, inc. SunGridEngine, OpenGridEngine and similar.

**Comment by mikej888** [comment]
The VMWare VM I use for RiboViz development has 8GB memory, 4 processors, 40GB disk (running Ubuntu 18.04 OS).

**Comment by XUEXUEXUE0** [comment]
[Run riboviz on Eddie.pdf](https://github.com/riboviz/riboviz/files/4676202/Run.riboviz.on.Eddie.pdf)
I wrote a pdf file which contains a hello nextflow example on Eddie and where I stuck. Could anyone  read it please and give me some suggestions?

**Comment by mikej888** [comment]
The Nextflow version I use is 20.01.0 from February 2020. The current version is 20.04.1 released 22 days ago...

https://github.com/nextflow-io/nextflow/releases/tag/v20.04.1.

From your log, the version on Eddie seems to be 0.26.0,

https://github.com/nextflow-io/nextflow/releases/tag/v0.26.0

from November 2017. This is very very out of date and is not even a v1.0 first version release. Nextflow have made myriad improvements and extensions and changes in the last 2.5 years which will explain the syntax errors.

A first step could be to find out if Eddie supports a more recent version of Nextflow (a more recent version of the module).

A second step could be to ask if the newest version can be installed of Eddie. There is little benefit, and a lot drawbacks in only offering such an out of date prototype version (it prevents users from benefiting from improvements, bug fixes, enhancements and from being able to run others' workflows).

**Comment by ewallace** [comment]
Sounds good. @XUEXUEXUE0 , could you find out if Eddie supports a more recent nextflow version? What if any help do you need with that?

**Comment by mikej888** [comment]
Try:

```console
module load anaconda
conda install nextflow
conda install -y -c bioconda cutadapt 
```

**Comment by mikej888** [comment]
I found https://www.wiki.ed.ac.uk/display/ResearchServices/Anaconda which describes how to use conda on EDDIE.

**Comment by ewallace** [comment]
Following @mikej888 's comment above, our goal is to have a riboviz 2.0 conda environment setup for bio_wallace_rna_riboviz

`/exports/<COLLEGE>/eddie/<SCHOOL>/groups/<GROUP NAME>/anaconda/pkgs`
I think this resolves to:
`/exports/cse/eddie/bio/groups/wallace_rna/anaconda/pkgs`

**Comment by mikej888** [comment]
If it's not `cse` it might be `scse`.

**Comment by XUEXUEXUE0** [comment]
Hi all, when I run the command line, it says
`cd /exports/scse/eddie/bio/groups/wallace_rna`
`-bash: cd: /exports/scse/eddie/bio/groups/wallace_rna: No such file or directory`
`cd /exports/cse/eddie/bio/groups/wallace_rna`
`-bash: cd: /exports/cse/eddie/bio/groups/wallace_rna: No such file or directory`

Is it because I don't have permission ?

**Comment by XUEXUEXUE0** [comment]
I have created a branch called `run-on-eddie-59` and put the documentation about how to run riboviz on Eddie in the `docs` folder. Comments/changes welcome.

**Comment by mikej888** [comment]
I've run following your instructions and it worked :-) (the biggest issue was trying to get access to an interactive node with 8GB!)

I had to create a `.condarc` file and use the R packages required by RiboViz that I'd installed in my home directory.

Rather than running in the copy of the `riboviz` repository in `/exports/csce/eddie/biology/groups/wallace_rna/`, I ran it in a clone of the `riboviz` repository, in my home directory, so I wouldn't overwrite any of your files.

Here are some suggestions for the documentation:

* "Troubleshooting: fail to enter inside node". Typo: inside => interactive
* Add that another reason this can fail is that EDDIE is running OK but there are no free nodes at present.
* Add that the user needs to have a `.condarc` file configured to point to the `anaconda` directory in `/exports/csce/eddie/biology/groups/wallace_rna/`:

```
envs_dirs:
  - /exports/csce/eddie/biology/groups/wallace_rna/anaconda/envs
pkgs_dirs:
  - /exports/csce/eddie/biology/groups/wallace_rna/anaconda/pkgs
```

* Add that `/exports/csce/eddie/biology/groups/wallace_rna` has Anaconda packages (in `anaconda`) and all the Python packages required by RiboViz are there, accessible as a `riboviz` conda environment.
* Add that the user needs to have installed all the R packages required by RiboViz, in an `R` directory in their home directory.
* Add that the `module load` and `source activate` commands can be put into a script, for example, `set-riboviz-env.sh`, in `/exports/csce/eddie/biology/groups/wallace_rna/`:

```
module load igmm/apps/BEDTools 
module load igmm/apps/bowtie
module load igmm/apps/hdf5
module load igmm/apps/HISAT2
module load igmm/apps/pigz
module load R/3.5.3
module load anaconda
source activate riboviz
```

* Then, in future, a user needs only to run:

```console
$ source set-riboviz-env.sh
```

* Add that regression test data, https://github.com/riboviz/regression-test-data-2.0.beta, has been cloned into `/exports/csce/eddie/biology/groups/wallace_rna/`.
* Move the instructions to `docs/user/` subdirectory.

**Comment by XUEXUEXUE0** [comment]
I installed R 3.6.3 packages in `/exports/csce/eddie/biology/groups/wallace_rna/Rlibrary` so users do not need to  install the R packages in their `home` directory. There is a huge change in my documentation. 
Do I need to add `$ pytest riboviz/test/regression/test_regression.py --expected=$HOME/regression-test-data-2.0.beta/` stuff in my documentation?

**Comment by kavousan** [comment]
Hello,

Not sure if you were aware, but Eddie will be at risk 09:00 AM, 29-Jun-20 — 05:00 PM, 17-Jul-20. They have no plan to take the system down, but they may have to, or the system may become unresponsive at times etc.

https://alerts.is.ed.ac.uk/index.cfm?fuseaction=view_alert&alert_id=9389

**Comment by kavousan** [comment]
To review and try out new documentation, and once happy, we close this ticket.

**Comment by ewallace** [comment]
I'm trying this out. Within 15 minutes I was running nextflow riboviz on an Eddie interactive node, so the documentation works!

Some comments:
- Title of `run-on-Eddie.md` should indicate this is for U. of Edinburgh users. 
- `run-on-Eddie.md` could link to Eddie homepage.
- `run-on-Eddie.md` should state that these guidelines _may_ be useful for running riboviz in other HPC systems.
- What does `activate` mean? Then when the command prompt starts with `(riboviz)` - is that a conda environment?
- The script `set-riboviz-env.sh` is great! Can we mention this earlier in the doc and link down to that heading?
- for `git clone https://github.com/riboviz/riboviz`, I suggest `mkdir riboviz` then `cd riboviz` first. That way we're in `home/riboviz/riboviz`, and the `example-datasets` repo can be put in `home/riboviz/example-datasets`. Any comments?

**Comment by FlicAnderson** [comment]
I've got nextflow riboviz and python riboviz running on Eddie in the interactive node too, the documentation was helpful.  

I've got a few suggestions too, but definitely agree with the things @ewallace has said [above](https://github.com/riboviz/riboviz/issues/59#issuecomment-651910175). 

Worth adding a little more information on how to adapt the process to run full-size datasets, and make a little clearer what parts of the documentation need to be followed only once, or which can be skipped if they are already setup for other users in the group? 

*Interactive sessions*: do we actually need to login to an interactive session in order to run riboviz? If we're using qsub and job submission scripts, we shouldn't need this, so might be worth removing this step of the documentation, or making it very clear that we don't want to try running riboviz this way normally? 

*Directory location*: I think it would be good to add information on which directory location is best to run different datasets from. For example it's not wise to run any full-size datasets from my home directory, which means I need to move to the lab data storage location.  I'm also not 100% clear on how you've negotiated upload/download/storage of full-size data files like the Cryptococcus data @XUEXUEXUE0 ? Are there any things to be aware about that? What do you recommend on this?

I wasn't sure whether I needed to run all the steps in the document, or if these necessary settings and setup steps had already been applied to the Wallace lab group by @XUEXUEXUE0 and I just had to navigate and run to/from there? Or in fact whether I could run from my home directory with the environment setup script? *Can you make clearer what a new user from our group would need to do to be able to run a) the vignette and b) a full-size dataset on Eddie? and then if they need to do all of those steps again the second time they log in to run something?*  I second @ewallace 's suggestion to mention the super environment setup script earlier in the doc.

Maybe also worth noting the *anaconda load warning*:
```
$ module load anaconda
WARNING: If you use conda to create environments, your home directory may fill up. Please see our documentation at 
 https://www.wiki.ed.ac.uk/display/ResearchServices/Anaconda for advice.
```
Does the `.condarc` file configuration avoid this problem?  Worth mentioning that your idea will stop that happening if so :)

Perhaps it's also good to *take out the run vignette interactively* parts of the documentation, as Mike & Edward mentioned [above](https://github.com/riboviz/riboviz/issues/59#issuecomment-642112783)?  Or possibly leave it in but make it clear that this is just an optional step to check riboviz is set up right. I didn't realise while following this documentation that this wasn't advised, so best to take those rows out, and jump straight to the job submission instructions instead.

**Comment by FlicAnderson** [comment]
Regarding this from @ewallace : 
> * for `git clone https://github.com/riboviz/riboviz`, I suggest `mkdir riboviz` then `cd riboviz` first. That way we're in `home/riboviz/riboviz`, and the `example-datasets` repo can be put in `home/riboviz/example-datasets`. Any comments?
I think I agree, but does this apply to the wallace_rna lab data area?

**Comment by FlicAnderson** [comment]
Another question re: wallace_rna/riboviz folder via Eddie: how does git handle being on Eddie and multiple users changing branches/etc?  If I login, change to that directory and run git status, and I find it's in run-on-eddie-59, if I run git checkout develop, does that alter it to develop for anyone else running on Eddie? Does it affect other jobs? Can we maybe clarify this point just so @XUEXUEXUE0 and I can be sure our work won't cause issues for each other if we're working on Eddie at the same time?

**Comment by FlicAnderson** [comment]
And also, it would be good to add information about how to submit and run jobs in the **priority queue** for Eddie users, since this is something you'll be doing?

Currently I've got a job to run the vignette waiting in the normal queue. Wonder how long it'll take to get run :smile:

**Comment by mikej888** [comment]
> Another question re: wallace_rna/riboviz folder via Eddie: how does git handle being on Eddie and multiple users changing branches/etc? If I login, change to that directory and run git status, and I find it's in run-on-eddie-59, if I run git checkout develop, does that alter it to develop for anyone else running on Eddie? Does it affect other jobs? Can we maybe clarify this point just so @XUEXUEXUE0 and I can be sure our work won't cause issues for each other if we're working on Eddie at the same time?

Having multiple users writing configuration files using the same clone of riboviz may be too error prone especially if the users need to tweak the code too. Options:

1. Users clone riboviz into their home directories but use the shared Python/R libraries in wallace_rna/riboviz. They can also put any large data files into wallace_rna/riboviz and configure their riboviz accordingly.
2. Users clone a personal copy of riboviz into the wallace_rna/riboviz folder. I originally did this (riboviz-mikej).

**Comment by XUEXUEXUE0** [comment]
I agree with @mikej888 

Obvioulsy, the space in home directory is not enough for running a full-size example dataset. So I ran it on my scratch space using the shared Python/R libararies in wallace_rna/riboviz. Even though the files there will be deleted after one month. But before that I will transfer important data elsewhere

I don't think most users have access to the group folder. So it is better to run a full-size example dateset on their scratch space?

**Comment by XUEXUEXUE0** [comment]
>And also, it would be good to add information about how to submit and run jobs in the priority queue for Eddie users, since this is something you'll be doing?

Priority queue is not free and Edward has paid the charge for us. 

To use the priority queue:

`qsub -P bio_wallace_rna_riboviz [Your script]`

But I don't think other users who are not our group members could use this prority queue

**Comment by XUEXUEXUE0** [comment]
Finished my third version doc! Comments/changes welcome!

**Comment by mikej888** [comment]
@XUEXUEXUE0 The latest doc tells users to install the latest version of the SRA toolkit, following https://github.com/ncbi/sra-tools/wiki/02.-Installing-SRA-Toolkit. This recommends a user does:

```console
$ vdb-config -i
```

If they don't do this then the commands won't run e.g.

```console
$ fasterq-dump --stdout SRR390728 | head -n 8
This sra toolkit installation has not been configured.
Before continuing, please run: vdb-config --interactive
For more information, see https://www.ncbi.nlm.nih.gov/sra/docs/sra-cloud/
```

If a user runs

```console
$ vdb-config -i
```

then they get a configuration window:

<img width="791" alt="sratoolkit" src="https://user-images.githubusercontent.com/2621549/86902134-74fdaa80-c105-11ea-8a37-6cfd2ad99f57.png">

What settings do they need to provide to enable the tools?

**Comment by XUEXUEXUE0** [comment]
Hi @mikej888 , my configuration window format is a little bit different from you. This is my setting:

![image](https://user-images.githubusercontent.com/55735374/86928738-8aa3bc00-c167-11ea-9e7e-b609fdd7419b.png)

![image](https://user-images.githubusercontent.com/55735374/86928809-a27b4000-c167-11ea-91a5-e5a9753dea3a.png)

And the rest setting is default

**Comment by mikej888** [comment]
@XUEXUEXUE0 thanks. I'd configured my PATH incorrectly so the EDDIE installation of vdb-config was being run and not the copy in the SRA Toolkit that I'd downloaded. I now see the same interface as you and fasterq-dump doesn't raise an error.

**Comment by ewallace** [comment]
I will test the updated documentation covering `fasterq-dump` and the Wallace 2020 Cryptococcus dataset. When that pans out, I expect that we should close this issue.

**Comment by ewallace** [comment]
I emailed ECDF (is.helpline) to ask them to install sratoolkit v 2.9.1 that includes the `fasterq-dump` command. That seems like a better way to do it.

Aspera connect is installed: `igmm/apps/asperaconnect/3.7.2` is one result of `module available` on Eddie.

When those are installed, we can update the docs accordingly.

**Comment by ewallace** [comment]
Ewan at ECDF installed SRA toolkit v2.10.8 to `igmm/apps/sratoolkit/2.10.8`. I've tested that `prefetch` and `fasterq-dump` work, and updated the documentation in `run-on-Eddie.md` in commit 11ecf12. Including information setting up the cache directory for SRA toolkit to point to Eddie scratch space.

This SRA download is a multi-step hassle. Apparently European Nucleotide Archive (ENA) allows you to directly download fastq.gz format SRA data, on ftp://ftp.sra.ebi.ac.uk/. Would that download be more efficient or at least take fewer steps?

**Comment by FlicAnderson** [comment]
I'm not sure if this is the best location for this question, but do the SRA.fastq files downloaded via the SRA toolkit method (e.g. after running `fasterq-dump SRR9620586` to produce SRR9620586.fastq) need to be zipped up with `pigz`? Or can riboviz handle .fastq read files instead of .fastq.gz files? This is advised in [run-on-Eddie.md](https://github.com/riboviz/riboviz/blob/run-on-eddie-59/docs/user/run-on-Eddie.md#download-fastq-data-files-from-the-short-read-archive-sra)

If I'm working on Eddie (scratch), and space is not necessarily such a concern, is running `pigz *.fastq` necessary, or is there a part of the pipeline which unzips the .fastq.gz files early on? 

@ewallace and @mikej888 is this something either of you know offhand?  Is it done this way just to create more running space on local machines?

**Comment by ewallace** [comment]
I *think* that riboviz accepts either `.fastq` or `fastq.gz`, because the first step is `cutadapt`, which [accepts compressed or uncompressed files as input](https://cutadapt.readthedocs.io/en/stable/guide.html). However I haven't tested that. So it shouldn't matter in principle whether `fasterq-dump` outputs compressed or not, but in practice it might matter. 

Is it possible to pipe the output of `fasterq-dump` directly into `pigz` for compression?

**Comment by FlicAnderson** [comment]
Advice on resource allocation from Mike Wallis on how to assign resource allocation on Eddie:

> Riboviz is an interesting one; although you can specify nproc I think it’s possible to overrequest and that causes some interesting issues with Eddie. Anyway, let’s answer your questions. 

> A decent “starter” for parallel would be to request a similar number of resources as you’d use on your laptop, but you might find it runs faster on Eddie. If I remember correctly i9s have up to 8 real cpus but allow hyperthreading; we don’t allow hyperthreading, so requesting 16 cpus would get you 16 cpus.

> As far as memory allocation goes you can request 4Gb per core without any penalty as that’s the minimum we have on any single node. It’s possible to request 16Gb/core but there’s fewer of those machines, so it’s a balancing act between how quickly you want to wait for your job to run and how much RAM you need.

> Try this:

 > #$ -l h_vmem=4G -pe sharedmem=16

> This will give you 64Gb RAM and 16 CPUs, so you can specify nprocesses in your code to 16 – you can then do some benchmarking between Eddie and your laptop (and be able to use your laptop while the workload is processed!) and make any changes as appropriate. Please don’t set nprocesses higher than the number of requested cores, as we have an alert when a job uses more process threads than requested slots (it usually means something has gone wrong) and it’ll affect all the processes running on that node.

> One thing to look out for: when your job is complete, have a look at the memory usage stats that you can get out of “qacct –j <job_id>” – this will tell you if you needed all the memory that you requested, or tell you that it tried to use more than you requested.

**Comment by FlicAnderson** [comment]
> Advice on resource allocation from Mike Wallis on how to assign resource allocation on Eddie:
> 
> > Riboviz is an interesting one; although you can specify nproc I think it’s possible to overrequest and that causes some interesting issues with Eddie. Anyway, let’s answer your questions.
> 
> > A decent “starter” for parallel would be to request a similar number of resources as you’d use on your laptop, but you might find it runs faster on Eddie. If I remember correctly i9s have up to 8 real cpus but allow hyperthreading; we don’t allow hyperthreading, so requesting 16 cpus would get you 16 cpus.
> 
> > As far as memory allocation goes you can request 4Gb per core without any penalty as that’s the minimum we have on any single node. It’s possible to request 16Gb/core but there’s fewer of those machines, so it’s a balancing act between how quickly you want to wait for your job to run and how much RAM you need.
> 
> > Try this:
> 
> > #$ -l h_vmem=4G -pe sharedmem=16
> 
> > This will give you 64Gb RAM and 16 CPUs, so you can specify nprocesses in your code to 16 – you can then do some benchmarking between Eddie and your laptop (and be able to use your laptop while the workload is processed!) and make any changes as appropriate. Please don’t set nprocesses higher than the number of requested cores, as we have an alert when a job uses more process threads than requested slots (it usually means something has gone wrong) and it’ll affect all the processes running on that node.
> 
> > One thing to look out for: when your job is complete, have a look at the memory usage stats that you can get out of “qacct –j <job_id>” – this will tell you if you needed all the memory that you requested, or tell you that it tried to use more than you requested.

I'll add this information into the run-on-eddie.md documentation, and also test how it goes with the Cryptococcus example-datasets/#3 and example-datasets/#4  data I'm working on getting through Eddie just now.

**Comment by FlicAnderson** [comment]
I will run a full-size dataset through this, update the documentation and close this ticket ASAP; aiming for closing by the weekend.

@ewallace suggestion to run multiple submissions of the same job with different parameters of resources at the same time, to get a better idea of what the 'best' numbers are to get things working. Good ideal!

**Comment by ewallace** [comment]
Given that multiple users have successfully run riboviz on Eddie, this issue is ready to review and close, and documentation committed to `develop` branch. Any remaining questions need to be divided up and created in separate issue tickets.

**Comment by FlicAnderson** [comment]
Will add documentation to the branch and commit by end of this week (ie by 02 Oct). 

I've got multiple datasets run on Eddie (vignette, Cryptococcus H99, JEC21) and have just submitted Brar 2012 dataset with 6 samples.  It's given me a much better idea of how it all works and the processes required for getting things running. I'll add those things in, obviously.

I'll be adding two example Eddie job-submission scripts (vignette + 1 'full-size' dataset example matching an example-datasets dataset) to `riboviz/bash` for #217 which I can link into the documentation at that point.

I've also figured out how best to set up the system links to have the input .fastq files and the output files held on the user's scratch space while running riboviz from their home directory.

**Comment by FlicAnderson** [comment]
I've updated the documentation in `/docs/user/run-on-eddie.md` for the JEC21 example dataset instructions section.  I still need to check over the initial parts, although I've also edited the vignette run to use nextflow command instead of python. 

I'll check this over again tomorrow around lunchtime, and should be able to PR before tomorrow's dev meeting.

**Comment by FlicAnderson** [comment]
Realised that I hadn't merged `develop` into my local copy of `run-on-eddie-59` for a while, so I did that, and I have a merge conflict to resolve in `run-on-eddie.md` :D  I have now fixed the conflict, and merged eleventy-billion commits from develop in...

Always pull first before editing, kids!
